### ISSUE TITLE
grpo  Async mode nccl timeout

### ISSUE THREAD

**Issue by xygnms66** [modelscope/ms-swift#4306]
**Describe the bug**
一卡进行部署
CUDA_VISIBLE_DEVICES=0 \
swift rollout 
三卡进行训练
swift rlhf \
    --rlhf_type grpo \
报错
rank1]:[E521 19:04:22.127741941 ProcessGroupNCCL.cpp:616] [Rank 1] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=131, OpType=_ALLGATHER_BASE, NumelIn=8, NumelOut=24, Timeout(ms)=1800000) ran for 1800022 milliseconds before timing out.
[rank1]:[E521 19:04:22.155606022 ProcessGroupNCCL.cpp:1785] [PG ID 0 PG GUID 0(default_pg) Rank 1] Exception (either an error or timeout) detected by watchdog at work: 131, last enqueued NCCL work: 131, last completed NCCL work: 130.
[rank1]:[E521 19:04:23.802884296 ProcessGroupNCCL.cpp:1834] [PG ID 0 PG GUID 0(default_pg) Rank 1] Timeout at NCCL work: 131, last enqueued NCCL work: 131, last completed NCCL work: 130.
[rank1]:[E521 19:04:23.802922636 ProcessGroupNCCL.cpp:630] [Rank 1] Some NCCL operations have failed or timed out. Due to the asynchronous nature of CUDA kernels, subsequent GPU operations might run on corrupted/incomplete data.
[rank1]:[E521 19:04:23.802936264 ProcessGroupNCCL.cpp:636] [Rank 1] To avoid data inconsistency, we are taking the entire process down.
[rank1]:[E521 19:04:23.818448707 ProcessGroupNCCL.cpp:1595] [PG ID 0 PG GUID 0(default_pg) Rank 1] Process group watchdog thread terminated with exception: [Rank 1] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=131, OpType=_ALLGATHER_BASE, NumelIn=8, NumelOut=24, Timeout(ms)=1800000) ran for 1800022 milliseconds before timing out.
Exception raised from checkTimeout at ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:618 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x96 (0x7f3eef16c446 in /mnt/nas/home/lxy/miniconda3/envs/swift/lib/python3.10/site-packages/torch/lib/libc10.so)
frame #1: c10d::ProcessGroupNCCL::WorkNCCL::checkTimeout(std::optional<std::chrono::duration<long, std::ratio<1l, 1000l> > >) + 0x282 (0x7f3ea502a772 in /mnt/nas/home/lxy/miniconda3/envs/swift/lib/python3.10/site-packages/torch/lib/libtorch_cuda.so)
frame #2: c10d::ProcessGroupNCCL::watchdogHandler() + 0x233 (0x7f3ea5031bb3 in /mnt/nas/home/lxy/miniconda3/envs/swift/lib/python3.10/site-packages/torch/lib/libtorch_cuda.so)
frame #3: c10d::ProcessGroupNCCL::ncclCommWatchdog() + 0x14d (0x7f3ea503361d in /mnt/nas/home/lxy/miniconda3/envs/swift/lib/python3.10/site-packages/torch/lib/libtorch_cuda.so)
frame #4: <unknown function> + 0x145c0 (0x7f3eef5f05c0 in /mnt/nas/home/lxy/miniconda3/envs/swift/lib/python3.10/site-packages/torch/lib/libtorch.so)
frame #5: <unknown function> + 0x94ac3 (0x7f3ef00e9ac3 in /lib/x86_64-linux-gnu/libc.so.6)
frame #6: <unknown function> + 0x126850 (0x7f3ef017b850 in /lib/x86_64-linux-gnu/libc.so.6)

**Your hardware and system info**
Write your system info like CUDA version/system/GPU/torch version here(在这里给出硬件信息和系统信息，如CUDA版本，系统，GPU型号和torch版本等)

**Additional context**
Add any other context about the problem here(在这里补充其他信息)

**Comment by hjh0119** [comment]
Check if there are any errors on the server and whether the host and port are correct.

**Comment by hjh0119** [comment]
Please try the latest main branch to see if the issue occurs there as well. If it does, please provide a script to reproduce the problem.

**Comment by hjh0119** [comment]
Can't reproduce in the latest code; it may be due to long inference times after setting dynamic_sample. Plz try using a smaller model and reducing the batch size.

**Comment by hjh0119** [comment]
I also used multi-GPU training, but can't reproduce the issue. The script is as follows:

```
CUDA_VISIBLE_DEVICES=6,7 \
swift rollout --model Qwen/Qwen2.5-3B-Instruct --data_paralllel_size 2
```

```
CUDA_VISIBLE_VICES=0,1,2,3,4,5 \
NPROC_PER_NODE=6 \
swift rlhf \
--rlhf_type grpo \
--model Qwen/Qwen2.5-3B-Instruct \
--train_type lora \
--dataset AI-MO/NuminaMath-TIR#10000 \
--torch_dtype bfloat16 \
--num_train_epochs 1 \
--vllm_server_host 127.0.0.1 \
--per_device_train_batch_size 2 \
--per_device_eval_batch_size 2 \
--gradient_accumulation_steps 4 \
--split_dataset_ratio 0 \
--save_steps 1000 \
--learning_rate 1e-6 \
--save_total_limit 2 \
--logging_steps 1 \
--output_dir output \
--warmup_ratio 0.05 \
--dataloader_num_workers 4 \
--max_completion_length 2048 \
--reward_funcs accuracy format \
--reward_weights 1 0.1 \
--num_generations 4 \
--system swift/examples/train/grpo/prompt.txt \
--use_vllm true \
--deepspeed zero2 \
--dynamic_sample true \
--overlong_filter true \
--temperature 1.0 \
--top_p 1.0 \
--top_k 80 \
--log_completions true \
--sleep_level 1 \
--max_steps 50 
```

**Comment by KevinClaint** [comment]
@hjh0119 I meet this bug again in version 3.7 and I tried for many times. It always occur in specific steps even if i train the model by  resuming from  checkpoint. This bug always occur after 3000steps.  The first i meet is 3770 and the second is 4320.

**Comment by KevinClaint** [comment]
CUDA_VISIBLE_DEVICES=4,7 \
swift rollout \
  --model /sharefs/model/QwenVL2.5_3B/ \
  --model_type qwen2_5_vl \
  --tensor_parallel_size 2 \
  --data_parallel_size 1 \
  --gpu_memory_utilization 0.9 \
  --max_model_len 4096 


MASTER_ADDR=localhost \
MASTER_PORT=29601 \
MAX_PIXELS=1003520 \
CUDA_VISIBLE_DEVICES=0,1,2,3 \
NPROC_PER_NODE=4 \
swift rlhf \
    --rlhf_type grpo \
    --model /sharefs/model/QwenVL2.5_3B/  \
    --model_type qwen2_5_vl \
    --external_plugins /home/liver01/lyz/ms-swift-3.6/examples/train/grpo/plugin/plugin.py \
    --reward_funcs xxx \
    --reward_weights 1 \
    --use_vllm true \
    --vllm_mode server \
    --vllm_server_host 127.0.0.1 \
    --vllm_server_port 8000 \
    --train_type full \
    --torch_dtype bfloat16 \
    --dataset /sharefs/data/grpo_format.jsonl \
    --max_completion_length 1024 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 2 \
    --learning_rate 1e-6 \
    --eval_steps 2000 \
    --save_steps 2000 \
    --save_total_limit 3 \
    --logging_steps 5 \
    --output_dir /sharefs/model/grpo \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --dataset_num_proc 4 \
    --num_generations 8 \
    --temperature 0.7 \
    --top_p 0.9 \
    --top_k 50 \
    --async_generate false \
    --system '/home/ms-swift/examples/train/grpo/prompt.txt' \
    --deepspeed zero2 \
    --log_completions true \
    --num_iterations 1 \
    --report_to tensorboard 

my shell is above

below is my environment:

absl-py==2.3.1
accelerate==1.8.1
addict==2.4.0
aiofiles==24.1.0
aiohappyeyeballs==2.6.1
aiohttp==3.12.13
aiosignal==1.4.0
airportsdata==20250706
aliyun-python-sdk-core==2.16.0
aliyun-python-sdk-kms==2.16.5
annotated-types==0.7.0
antlr4-python3-runtime==4.13.2
anyio==4.9.0
astor==0.8.1
async-timeout==5.0.1
attrdict==2.0.1
attrs==25.3.0
av==15.0.0
binpacking==1.5.2
blake3==1.0.5
cachetools==6.1.0
certifi==2025.7.9
cffi==1.17.1
charset-normalizer==3.4.2
click==8.2.1
cloudpickle==3.1.1
compressed-tensors==0.9.3
contourpy==1.3.2
cpm-kernels==1.0.11
crcmod==1.7
cryptography==45.0.5
cupy-cuda12x==13.4.1
cycler==0.12.1
dacite==1.9.2
datasets==3.3.2
deepspeed==0.16.9
Deprecated==1.2.18
depyf==0.18.0
dill==0.3.8
diskcache==5.6.3
distro==1.9.0
dnspython==2.7.0
einops==0.8.1
email_validator==2.2.0
exceptiongroup==1.3.0
fastapi==0.116.0
fastapi-cli==0.0.8
fastapi-cloud-cli==0.1.2
fastrlock==0.8.3
ffmpy==0.6.0
filelock==3.18.0
flash_attn @ file:///home/flash_attn-2.7.4.post1%2Bcu12torch2.6cxx11abiFALSE-cp310-cp310-linux_x86_64.whl#sha256=ffe17686fa1a0f288de9eae7c32af209d32a27b037ef28614f042b377af5b15a
fonttools==4.58.5
frozenlist==1.7.0
fsspec==2024.12.0
future==1.0.0
gguf==0.17.1
googleapis-common-protos==1.70.0
gradio==5.35.0
gradio_client==1.10.4
groovy==0.1.2
grpcio==1.73.1
h11==0.16.0
hf-xet==1.1.5
hjson==3.1.0
httpcore==1.0.9
httptools==0.6.4
httpx==0.28.1
huggingface-hub==0.33.2
idna==3.10
importlib_metadata==8.0.0
iniconfig==2.1.0
interegular==0.3.3
jieba==0.42.1
Jinja2==3.1.6
jiter==0.10.0
jmespath==0.10.0
joblib==1.5.1
jsonschema==4.24.0
jsonschema-specifications==2025.4.1
kiwisolver==1.4.8
lark==1.2.2
latex2sympy2_extended==1.10.2
llguidance==0.7.30
llvmlite==0.44.0
lm-format-enforcer==0.10.11
Markdown==3.8.2
markdown-it-py==3.0.0
MarkupSafe==3.0.2
math-verify==0.8.0
matplotlib==3.10.3
mdurl==0.1.2
mistral_common==1.6.3
modelscope==1.28.0
mpmath==1.3.0
Editable install with no version control (ms_swift==3.7.0.dev0)
-e /home/ms-swift
msgpack==1.1.1
msgspec==0.19.0
multidict==6.6.3
multiprocess==0.70.16
nest-asyncio==1.6.0
networkx==3.4.2
ninja==1.11.1.4
nltk==3.9.1
numba==0.61.2
numpy==1.26.4
nvidia-cublas-cu12==12.4.5.8
nvidia-cuda-cupti-cu12==12.4.127
nvidia-cuda-nvrtc-cu12==12.4.127
nvidia-cuda-runtime-cu12==12.4.127
nvidia-cudnn-cu12==9.1.0.70
nvidia-cufft-cu12==11.2.1.3
nvidia-cufile-cu12==1.11.1.6
nvidia-curand-cu12==10.3.5.147
nvidia-cusolver-cu12==11.6.1.9
nvidia-cusparse-cu12==12.3.1.170
nvidia-cusparselt-cu12==0.6.2
nvidia-ml-py==12.575.51
nvidia-nccl-cu12==2.21.5
nvidia-nvjitlink-cu12==12.4.127
nvidia-nvtx-cu12==12.4.127
nvitop==1.5.1
openai==1.93.2
opencv-python-headless==4.11.0.86
opentelemetry-api==1.26.0
opentelemetry-exporter-otlp==1.26.0
opentelemetry-exporter-otlp-proto-common==1.26.0
opentelemetry-exporter-otlp-proto-grpc==1.26.0
opentelemetry-exporter-otlp-proto-http==1.26.0
opentelemetry-proto==1.26.0
opentelemetry-sdk==1.26.0
opentelemetry-semantic-conventions==0.47b0
opentelemetry-semantic-conventions-ai==0.4.10
orjson==3.10.18
oss2==2.19.1
outlines==0.1.11
outlines_core==0.1.26
packaging==25.0
pandas==2.3.1
partial-json-parser==0.2.1.1.post6
peft==0.15.2
pillow==11.3.0
pluggy==1.6.0
prometheus-fastapi-instrumentator==7.1.0
prometheus_client==0.22.1
propcache==0.3.2
protobuf==4.25.8
psutil==7.0.0
py-cpuinfo==9.0.0
pyarrow==20.0.0
pybind11==2.13.6
pycountry==24.6.1
pycparser==2.22
pycryptodome==3.23.0
pydantic==2.11.7
pydantic_core==2.33.2
pydub==0.25.1
Pygments==2.19.2
pyparsing==3.2.3
pytest==8.4.1
python-dateutil==2.9.0.post0
python-dotenv==1.1.1
python-json-logger==3.3.0
python-multipart==0.0.20
pytz==2025.2
PyYAML==6.0.2
pyzmq==27.0.0
qwen-vl-utils==0.0.10
ray==2.47.1
referencing==0.36.2
regex==2024.11.6
requests==2.32.4
rich==14.0.0
rich-toolkit==0.14.8
rignore==0.5.1
rouge==1.0.1
rouge_score==0.1.2
rpds-py==0.26.0
ruff==0.12.2
safehttpx==0.1.6
safetensors==0.5.3
scikit-learn==1.7.0
scipy==1.15.3
semantic-version==2.10.0
sentence-transformers==5.0.0
sentencepiece==0.2.0
sentry-sdk==2.32.0
shellingham==1.5.4
simplejson==3.20.1
six==1.17.0
sniffio==1.3.1
sortedcontainers==2.4.0
starlette==0.46.2
sympy==1.13.1
tensorboard==2.19.0
tensorboard-data-server==0.7.2
threadpoolctl==3.6.0
tiktoken==0.9.0
tokenizers==0.21.2
tomli==2.2.1
tomlkit==0.13.3
torch==2.6.0
torchaudio==2.6.0
torchvision==0.21.0
tqdm==4.67.1
transformers==4.51.3
transformers-stream-generator==0.0.5
triton==3.2.0
trl==0.19.1
typer==0.16.0
typing-inspection==0.4.1
typing_extensions==4.14.1
tzdata==2025.2
urllib3==2.5.0
uvicorn==0.35.0
uvloop==0.21.0
vllm==0.8.5.post1
watchfiles==1.1.0
websockets==15.0.1
Werkzeug==3.1.3
wrapt==1.17.2
xformers==0.0.29.post2
xgrammar==0.1.18
xxhash==3.5.0
yarl==1.20.1
zipp==3.23.0
zstandard==0.23.0

**Comment by BoweiPu** [comment]
蹲
